inmem/kvlock.go

```diff
@@ -0,0 +1,127 @@
+package inmem
+
+import "context"
+
+//  This is an RWMutex using channels that allows for context timeouts.
+//  It is adapted from Rob Pike's blog post:
+//  https://blogtitle.github.io/go-advanced-concurrency-patterns-part-3-channels/
+// where he made an RWMutex using channels.  I just added the context handlng.
+// h/t to Bryan Mills for pointing out this blog post to me.
+type RWMutex struct {
+	write   chan struct{}
+	readers chan int
+}
+
+func NewLock() RWMutex {
+	return RWMutex{
+		// This is used as a normal Mutex.
+		write: make(chan struct{}, 1),
+		// This is used to protect the readers count.
+		// By receiving the value it is guaranteed that no
+		// other goroutine is changing it at the same time.
+		readers: make(chan int, 1),
```

Comment:
I think you're going to want to prime this channel with `readers <- 0` before returning.
Otherwise, the first attempt to read will block on the first retrieve.
This does get unblocked by the first write lock call, but I imagine we wouldn't want to rely on at least one Write before a Read can happen.

Reply:
Oh wait, I see you select on both and increment afterwards. Ignore me.